// @compository/lib/src/elements/discover-dnas.ts
import { LitElement, Constructor, css, html, property } from 'lit-element';
import { ScopedElementsMixin as Scoped } from '@open-wc/scoped-elements';
import { sharedStyles } from './sharedStyles';
import { Card } from 'scoped-material-components/mwc-card';
import { List } from 'scoped-material-components/mwc-list';
import { ListItem } from 'scoped-material-components/mwc-list-item';
import { Dictionary, serializeHash } from '@holochain-open-dev/core-types';
import { CircularProgress } from 'scoped-material-components/mwc-circular-progress';
import { Button } from 'scoped-material-components/mwc-button';
import { Snackbar } from 'scoped-material-components/mwc-snackbar';
import {
  CompositoryService,
  GetTemplateForDnaOutput,
} from '../services/compository-service';
import { generateDnaBundle } from '../processes/generate-dna-bundle';
import { InstallDnaDialog } from './install-dna-dialog';
import { BaseElement } from '@holochain-open-dev/common';
import { BaseCompositoryService } from './base';

export abstract class DiscoverDnas extends BaseCompositoryService {
  @property({ type: Boolean })
  _loading = true;

  @property({ type: Array })
  _allInstantiatedDnasHashes: Array<string> | undefined = undefined;
  // {[instantiated_dna_hash]: DnaTemplate}
  @property({ type: Array })
  _dnaTemplates: Dictionary<GetTemplateForDnaOutput> = {};

  async firstUpdated() {
    const cellIds = await this._compositoryService.adminWebsocket.listCellIds();
    const allInstalledDnaHashes = cellIds.map(cellId =>
      serializeHash(cellId[0])
    );
    const allInstantiatedDnas = await this._compositoryService.getAllInstantiatedDnas();
    this._allInstantiatedDnasHashes = allInstantiatedDnas.filter(
      hash => !allInstalledDnaHashes.includes(hash)
    );

    const promises = this._allInstantiatedDnasHashes.map(async hash => {
      const template = await this._compositoryService.getTemplateForDna(hash);
      return [hash, template] as [string, GetTemplateForDnaOutput];
    });

    const templateArray = await Promise.all(promises);

    for (const [hash, template] of templateArray) {
      this._dnaTemplates[hash] = template;
    }

    this._loading = false;
  }

  async displayInstallDna(dnaHash: string, retriesLeft: number = 3) {
    this._loading = true;

    if (retriesLeft === 0) {
      this._loading = false;
      return;
    }

    try {
      const template = this._dnaTemplates[dnaHash];

      const dnaBundle = await generateDnaBundle(
        this._compositoryService,
        template.dnaTemplate,
        template.properties,
        template.uuid
      );

      const dialog: InstallDnaDialog = this.shadowRoot?.getElementById(
        'install-dialog'
      ) as InstallDnaDialog;
      dialog.dnaBundle = dnaBundle;
      dialog.open();
    } catch (e) {
      (this.shadowRoot?.getElementById('error-snackbar') as Snackbar).show();
    }
  }

  renderContent() {
    return html` <mwc-card class="fill">
      <div class="column fill">
        <span style="margin: 16px; margin-bottom: 0;" class="title"
          >Discover DNAs</span
        >

        ${this._allInstantiatedDnasHashes?.length === 0
          ? html`
              <div class="fill column center-content">
                <span class="placeholder" style="max-width: 400px;"
                  >There are no DNAs generated by other agents that you don't
                  have installed</span
                >
              </div>
            `
          : html`
              <div class="flex-scrollable-parent">
                <div class="flex-scrollable-container">
                  <div class="flex-scrollable-y">
                    <mwc-list>
                      ${this._allInstantiatedDnasHashes?.map(
                        hash =>
                          html`
                            <div class="row" style="align-items: center;">
                              <mwc-list-item twoline style="flex: 1;">
                                <span
                                  >${this._dnaTemplates[hash].dnaTemplate
                                    .name}</span
                                >
                                <span slot="secondary">${hash}</span>
                              </mwc-list-item>
                              <mwc-button
                                label="INSTALL"
                                raised
                                @click=${() => this.displayInstallDna(hash)}
                                style="margin-right: 16px;"
                              ></mwc-button>
                            </div>
                          `
                      )}
                    </mwc-list>
                  </div>
                </div>
              </div>
            `}
      </div>
    </mwc-card>`;
  }

  renderErrorSnackbar() {
    return html`
      <mwc-snackbar
        id="error-snackbar"
        labelText="Couldn't generate the DNA due to gossip inconsistencies. Please try again in a few minutes."
      ></mwc-snackbar>
    `;
  }
  render() {
    return html`
      ${this.renderErrorSnackbar()}
      <install-dna-dialog id="install-dialog"></install-dna-dialog>
      ${this._loading
        ? html`<div class="fill center-content">
            <mwc-circular-progress indeterminate></mwc-circular-progress>
          </div>`
        : this.renderContent()}
    `;
  }

  static get styles() {
    return [
      sharedStyles,
      css`
        :host {
          display: flex;
        }
      `,
    ];
  }

  getScopedElements() {
    const compositoryService = this._compositoryService;
    return {
      'mwc-card': Card,
      'mwc-button': Button,
      'mwc-snackbar': Snackbar,
      'mwc-list': List,
      'mwc-circular-progress': CircularProgress,
      'mwc-list-item': ListItem,
      'install-dna-dialog': class extends InstallDnaDialog {
        get _compositoryService() {
          return compositoryService;
        }
      } as typeof HTMLElement,
    };
  }
}
